### PR TITLE
Pedrobn23/autoglome

### PR DESCRIPTION
**Pull request by pedrobn23**
This is a draft for modifications of PyGLOME. Still Work to do, will enjoy some feedback.

### CODE REVIEW

**Comment by pedrobn23** [comment]
PTAL.

File: python/pyglome/glome.py [hunk 1]
```diff
@@ -217,3 +217,143 @@ def generate_keys() -> KeyPair:
...
+    automatically generated making use of `generate_keys`. On initialization,
+    two counter (sending and receiving) are created and set to 0. Provides
+    methods tag (to generate new tags) and check (to check receiving tags).
+    """
+
+    def __init__(self,
+                 peer_key: x25519.X25519PublicKey,
+                 my_private_key: x25519.X25519PrivateKey = None,
+                 min_peer_tag_len: int = Glome.MAX_TAG_LEN,
+                 skippable_range: int = 0):
```
Comment (pedrobn23) [inline python/pyglome/glome.py:236]: Should these two arguments be param-only?

File: python/pyglome/glome.py [hunk 2]
```diff
@@ -217,3 +217,143 @@ def generate_keys() -> KeyPair:
...
+        Returns:
+           None.
+        Raises:
+           IncorrectTagError: Raised whenever the tag is incorrect.
+        """
+        old_counter = self._receiving_counter
+        for i in range(self.skippable_range + 1):
+            try:
+                self.glome.check(tag, msg, self.receiving_counter)
+                return None
```
Comment (burgerdev) [inline python/pyglome/glome.py:353]: Should there be an increment of the counter here?

File: python/test/autoglome_test.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        test_vector = self.test_vector
+        try:
+            test_vector.sender_glome.sending_counter = 0
+            test_vector.receiver_glome.sending_counter = 23
+            test_vector.sender_glome.receiving_counter = 123
+            test_vector.receiver_glome.receiving_counter = 255
+        except ValueError:
+            self.fail('properties raised ValueError unexpectedly!')
+
+    def test_check_counters_are_correctly_setted(self):
```
Comment (burgerdev) [inline python/test/autoglome_test.py:72]: typo: `..._correctly_set`

File: python/test/autoglome_test.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        test_vector.sender_glome.sending_counter = test_vector.data.counter
+        self.assertEqual(test_vector.sender_glome.tag(test_vector.data.msg),
+                         test_vector.data.tag)
+
+    def test_skippable_range(self):
+        test_vector = self.test_vector
+        try:
+            test_vector.receiver_glome.receiving_counter = test_vector.data.counter
+            test_vector.receiver_glome.check(test_vector.data.tag,
+                                             msg=test_vector.data.msg)
```
Comment (burgerdev) [inline python/test/autoglome_test.py:94]: I don't quite understand how this tests the skippable range, could you maybe describe what the test does?

File: python/test/test_vectors.py [hunk 5]
```diff
@@ -11,18 +11,36 @@
...
+    def __init__(self, kap: str, ka: str, kbp: str, kb: str, counter: int,
+                 msg: str, sk: str, tag: str):
+        """Constructor for TestVector Class."""
         self.kap = bytes.fromhex(kap)
         self.ka = bytes.fromhex(ka)
         self.kbp = bytes.fromhex(kbp)
         self.kb = bytes.fromhex(kb)
         self.counter = counter
-        self.msg = msg
+        self.msg = msg.encode()
```
Comment (burgerdev) [inline python/test/test_vectors.py:43]: Nit: can we do `.encode(encoding="ascii")`? It happens to be the same for our test vectors, but the spec demands ascii.

File: python/pyglome/glome.py [hunk 6]
```diff
@@ -217,3 +217,143 @@ def generate_keys() -> KeyPair:
...
+                 skippable_range: int = 0):
+        """Initialize AutoGlome class.
+
+        Performs the handshake, generates keys and counters.
+        Args:
+           peer_key: Your peer's public key.
+           my_private_key: Your private key.
+           min_peer_tag_len: Desired length (in bytes) for the tag.
+             Must be an integer in range 1-32.
+           skippable_range: Number of messages that can be missed.
```
[the remark this answers is not in the document]
Comment (burgerdev) [reply python/pyglome/glome.py:245]: Or maybe just refer to the `check` method's documentation.

File: python/pyglome/glome.py (hunk 1 above)
Comment (pedrobn23) [reply python/pyglome/glome.py:236]: Great!

File: python/pyglome/glome.py [hunk 7]
```diff
@@ -217,3 +217,143 @@ def generate_keys() -> KeyPair:
...
+    def sending_counter(self, value: int):
+        if not 0 <= value <= 255:
+            raise ValueError('Counter must be in range 0-255')
+        self._sending_counter = value
+
+    @property
+    def receiving_counter(self) -> int:
+        """Number of tags the user receives from the peer.
+
+        It is incremented each time a new tag is checked. It is
```
[the remark this answers is not in the document]
Comment (pedrobn23) [reply python/pyglome/glome.py:283]: It should says successfully checked. Fixed :).

File: python/pyglome/glome.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (pedrobn23) [reply python/pyglome/glome.py:245]: Added reference.

File: python/pyglome/glome.py (hunk 2 above)
Comment (pedrobn23) [reply python/pyglome/glome.py:353]: Yes, it should be.

File: python/test/test_vectors.py (hunk 5 above)
Comment (pedrobn23) [reply python/test/test_vectors.py:43]: Added :)

File: python/test/autoglome_test.py (hunk 3 above)
Comment (pedrobn23) [reply python/test/autoglome_test.py:72]: Fixed

File: python/test/autoglome_test.py (hunk 4 above)
Comment (pedrobn23) [reply python/test/autoglome_test.py:94]: It was wrong, and wasn't doing what I intended. It has been updated so now:
- set the receiving counter to counter-skippable range
- ensures the message is correctly checked
- assert the counter is correctly setted afterwards

File: python/pyglome/glome.py (hunk 2 above)
Comment (pedrobn23) [reply python/pyglome/glome.py:359]: fixed!

**Comment by pedrobn23** [comment]
PTAL

**Review by burgerdev — APPROVED** [review]
lgtm

**Review by pkern — CHANGES_REQUESTED** [review]
There's still some trailing whitespace in this change. Could we still fix that before we import it into Piper? The annoying bit is that neither Github nor Gerrit make that obvious, but it's super obvious in Critique. And I guess we should actually get pylint going as an additional CI check.

**Comment by pedrobn23** [comment]
Trailing whitespaces has been deleted, @pkern PTAL. Also, pyglome module has been checked against pylint3.

**Pull request merged**